design-proposals/amd-sev.md

```diff
@@ -0,0 +1,245 @@
+      <reducedPhysBits>1</reducedPhysBits>
+    </sev>
+    ...
+  </features>
+</domainCapabilities>
+```
+
+The `policy` parameter allows setting the guest policy flags. Additionally the optional `dhCert` and `session` provide the guest owner's base64 encoded `DH` (Diffie-Hellman) key and the guest owner's base64 encoded `launch blob` respectively.
+
+There are also several prerequisites to run SEV guests:
+
+- All virtio devices need to be configured with the `iommu='on'` attribute in their `<driver>` configuration
+- All memory regions used by the VM must be `locked` for Direct Memory Access (DMA) and to prevent swapping
+
+## Attestation
+
+[Attestation](https://github.com/AMDESE/sev-tool#proposed-provisioning-steps) is an interactive process between platform owner (`host`) and guest owner (`user`). It is carried out to establish a certain level of trust. To leverage the benefits of memory encryption a guest owner provides an encrypted VM disk. To decrypt the data and to run a VM with such a disk there is a need to `inject a secret` into the VM when the system is about to boot. The secret is known only by the guest owner. The goal of the attestation process is to validate that the platform is genuine and it really runs a non-compromised SEV hardware. Also it is needed to establish a private communication channel with the platform so the guest owner can securely provide the secret. Attestation consists of the following steps:
+
+### Step #1 Validation of the platform
+
+The guest owner should get the assets listed bellow from the platform owner:
+
+- the `Platform Diffie-Hellman (PDH)` key exported from the SEV hardware
+- the complete key chain (up to the `root of trust`) used to sign the `PDH`
+- the OVMF image
+
+On a node where KubeVirt is running, the privileged `virt-handler` pod can export the required assets in advance to a `kubevirt-private` directory. Whenever a new VM requiring SEV is launched, the directory can be mounted to the `virt-launcher` pod in `read-only` mode. A user can then invoke the `virtctl` tool to download the certificate chain for validation (for that an additional command will be needed, e.g. `virtctl fetchcertchain ...`).
+
+In order to ensure that the platform is not compromised the guest owner needs to validate the `PDH` key using the provided chain. For that `sev-tool --validate_cert_chain ...` can be used (AMD root certificate should be downloaded from the web for verification). Additionally the UEFI image should be considered 'trustable' by the guest owner.
+
+### Step #2 Establishing secure channel with the platform
+
+To establish a secure communication channel with the platform a guest owner needs to provide a launch blob with the parameters of the session and the certificate derived from `PDH`. This is currently not in the scope of KubeVirt as it requires the guest owner to expose the secrets. It should be performed 'offline' by the guest owner (e.g. by using `sev-tool --generate_launch_blob ...`).
```

Comment:
This part indeed sounds very tricky. Not sure how users can achieve this. 

What if the extraction of launch blob be done in an init container that would use a provided secret (I assume this is required if I read this correctly... ) and pass it to the virt launcher via env. variable?

Reply:
Yeah, maybe the explanation is a bit confusing. The launch blob is not extracted. Here the guest owner is responsible to geneate the blob. I think it is possible to pass certain parameters to virt-launcher but I dont think it will make things much easier. The procedure here implies also generation of several public/private key pairs and only the guest owner is supposed to have access to them (to private ones). If this task was to be offloaded to some init container then we would probably need to include it in the attestation as well.

The procedure is described in more details here https://github.com/enarx/sev/tree/main/docs/attestation#secure-channel-establishment. Its indeed tricky. In general I started thinking in the direction of having some external service for facilitating the attestation. Maybe KubeVirt could just provide the APIs to allow establishing this secure channel and injecting the secret afterwards. In such case the service would need to be 'trusted' by the guest owner but KubeVirt would in some sense stay away from the 'sensitive' data.